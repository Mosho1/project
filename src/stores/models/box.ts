import { types, getParent, hasParent } from 'mobx-state-tree'
import { pouch } from '../utils/pouchdb-model';
import { Socket, SocketTypeEnum } from './socket';
import { IStore } from '../domain-state';
import { CodeBlock } from './code-block';

interface BoxEditableProps {
    x?: number;
    y?: number;
    width?: number;
    height?: number;
    scaleX?: number;
    scaleY?: number;
}

export const Box = pouch.model('Box',
    {
        name: 'hal',
        x: 0,
        y: 0,
        width: 150,
        value: types.maybe(types.string),
        sockets: types.optional(types.array(Socket), []),
        code: types.reference(CodeBlock)
    })
    .views(self => ({
        get store(): IStore | null {
            if (!hasParent(self)) return null;
            return getParent(self, 2);
        },
        get inputs() {
            return self.sockets.filter(({ socketType }) => socketType === 'input');
        },
        get outputs() {
            return self.sockets.filter(({ socketType }) => socketType === 'output');
        },
        get execInputs() {
            return self.sockets.filter(({ socketType }) => socketType === 'execInput');
        },
        get execOutputs() {
            return self.sockets.filter(({ socketType }) => socketType === 'execOutput');
        },

    }))
    .views(self => ({
        get isSelected() {
            return Boolean(self.store && self.store.selection === self);
        },
        get height() {
            return 50 + Math.max(
                self.execInputs.length + self.inputs.length,
                self.execOutputs.length + self.outputs.length)
                * 30;
        }
    }))
    .actions(self => ({
        move(dx: number, dy: number) {
            self.x += dx;
            self.y += dy;
            return self;
        },
        setProps(props: BoxEditableProps) {
            /* istanbul ignore next */
            Object.assign(self, props);
        },
        addSocket(type: SocketTypeEnum, name = '') {
            const socket = Socket.create({ name: name, socketType: type });
            self.sockets.push(socket);
            return socket;
        },
        setValue(value: string) {
            self.value = value;
        }
    }));

type IBoxType = typeof Box.Type;
export interface IBox extends IBoxType { };
