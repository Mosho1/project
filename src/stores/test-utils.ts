import { mapValues } from 'lodash';

export const mock = <T>(instance: T, mock: NestedPartial<T>) => {
    for (let k in mock) {
        const descriptor = Object.getOwnPropertyDescriptor(instance, k);
        Object.defineProperties(instance, mapValues(mock, value => ({
            value,
            enumerable: descriptor!.enumerable,
            configurable: true,
            writable: true,
        })));
    }
    return instance;
};

const f = (a: any[], b: any[]): any[] =>
    [].concat(...a.map(a2 => b.map((b2: any) => [].concat(a2 as any, b2 as any) as any) as any));

export const product = (a: any[], b: any[], ...c: any[]): any[][] => {
    if (!b || b.length === 0) {
        return a;
    }
    const [b2, ...c2] = c;
    const fab = f(a, b);
    return product(fab, b2, ...c2);
};