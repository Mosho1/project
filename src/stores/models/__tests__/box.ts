import { Box } from '../box';
import { mock, product } from '../../test-utils';

test('height', () => {
    const lengths = [0, 1, 2];
    const tests = product(lengths, lengths, lengths, lengths);
    for (const test of tests) {
        expect(mock(Box.create(), {
            inputs: Array(test[0]),
            outputs: Array(test[1]),
            execInputs: Array(test[2]),
            execOutputs: Array(test[3]),
        }).height).toMatchSnapshot(test.join(','));
    }
});

test('isSelected', () => {
    expect(mock(Box.create(), {}).isSelected).toBe(false);
    expect(mock(Box.create(), { store: { selection: null } }).isSelected).toBe(false);
    const box = Box.create();
    expect(mock(box, { store: { selection: box } }).isSelected).toBe(true);
});

test('inputs', () => {
    expect(Box.create().inputs.length).toBe(0);
    expect(mock(Box.create(), { sockets: [{ socketType: 'input' }] }).inputs.length).toBe(1);
});

test('outputs', () => {
    expect(Box.create().outputs.length).toBe(0);
    expect(mock(Box.create(), { sockets: [{ socketType: 'output' }] }).outputs.length).toBe(1);
});

test('execInputs', () => {
    expect(Box.create().execInputs.length).toBe(0);
    expect(mock(Box.create(), { sockets: [{ socketType: 'execInput' }] }).execInputs.length).toBe(1);
});

test('execOutputs', () => {
    expect(Box.create().execOutputs.length).toBe(0);
    expect(mock(Box.create(), { sockets: [{ socketType: 'execOutput' }] }).execOutputs.length).toBe(1);
});

test('move', () => {
    const box = Box.create({ _id: 'test' });
    expect(box.move(1, 2)).toMatchSnapshot();
    expect(box.move(1, 2)).toMatchSnapshot();
});

test('addSocket', () => {
    const box = Box.create();
    expect(box.sockets).toHaveProperty('length', 0);
    box.addSocket('execInput')
    expect(box.sockets).toHaveProperty('length', 1);
    box.addSocket('input')
    expect(box.sockets).toHaveProperty('length', 2);
    expect(box.inputs).toHaveProperty('length', 1);
    expect(box.outputs).toHaveProperty('length', 0);
    expect(box.execInputs).toHaveProperty('length', 1);
    expect(box.execOutputs).toHaveProperty('length', 0);
});