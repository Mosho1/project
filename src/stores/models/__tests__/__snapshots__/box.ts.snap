// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`areSocketsCompatible: execInput,execInput,0,0 1`] = `false`;

exports[`areSocketsCompatible: execInput,execInput,0,1 1`] = `false`;

exports[`areSocketsCompatible: execInput,execInput,1,0 1`] = `false`;

exports[`areSocketsCompatible: execInput,execInput,1,1 1`] = `false`;

exports[`areSocketsCompatible: execInput,execOutput,0,0 1`] = `true`;

exports[`areSocketsCompatible: execInput,execOutput,0,1 1`] = `false`;

exports[`areSocketsCompatible: execInput,execOutput,1,0 1`] = `false`;

exports[`areSocketsCompatible: execInput,execOutput,1,1 1`] = `false`;

exports[`areSocketsCompatible: execInput,input,0,0 1`] = `false`;

exports[`areSocketsCompatible: execInput,input,0,1 1`] = `false`;

exports[`areSocketsCompatible: execInput,input,1,0 1`] = `false`;

exports[`areSocketsCompatible: execInput,input,1,1 1`] = `false`;

exports[`areSocketsCompatible: execInput,output,0,0 1`] = `false`;

exports[`areSocketsCompatible: execInput,output,0,1 1`] = `false`;

exports[`areSocketsCompatible: execInput,output,1,0 1`] = `false`;

exports[`areSocketsCompatible: execInput,output,1,1 1`] = `false`;

exports[`areSocketsCompatible: execOutput,execInput,0,0 1`] = `true`;

exports[`areSocketsCompatible: execOutput,execInput,0,1 1`] = `false`;

exports[`areSocketsCompatible: execOutput,execInput,1,0 1`] = `false`;

exports[`areSocketsCompatible: execOutput,execInput,1,1 1`] = `false`;

exports[`areSocketsCompatible: execOutput,execOutput,0,0 1`] = `false`;

exports[`areSocketsCompatible: execOutput,execOutput,0,1 1`] = `false`;

exports[`areSocketsCompatible: execOutput,execOutput,1,0 1`] = `false`;

exports[`areSocketsCompatible: execOutput,execOutput,1,1 1`] = `false`;

exports[`areSocketsCompatible: execOutput,input,0,0 1`] = `false`;

exports[`areSocketsCompatible: execOutput,input,0,1 1`] = `false`;

exports[`areSocketsCompatible: execOutput,input,1,0 1`] = `false`;

exports[`areSocketsCompatible: execOutput,input,1,1 1`] = `false`;

exports[`areSocketsCompatible: execOutput,output,0,0 1`] = `false`;

exports[`areSocketsCompatible: execOutput,output,0,1 1`] = `false`;

exports[`areSocketsCompatible: execOutput,output,1,0 1`] = `false`;

exports[`areSocketsCompatible: execOutput,output,1,1 1`] = `false`;

exports[`areSocketsCompatible: input,execInput,0,0 1`] = `false`;

exports[`areSocketsCompatible: input,execInput,0,1 1`] = `false`;

exports[`areSocketsCompatible: input,execInput,1,0 1`] = `false`;

exports[`areSocketsCompatible: input,execInput,1,1 1`] = `false`;

exports[`areSocketsCompatible: input,execOutput,0,0 1`] = `false`;

exports[`areSocketsCompatible: input,execOutput,0,1 1`] = `false`;

exports[`areSocketsCompatible: input,execOutput,1,0 1`] = `false`;

exports[`areSocketsCompatible: input,execOutput,1,1 1`] = `false`;

exports[`areSocketsCompatible: input,input,0,0 1`] = `false`;

exports[`areSocketsCompatible: input,input,0,1 1`] = `false`;

exports[`areSocketsCompatible: input,input,1,0 1`] = `false`;

exports[`areSocketsCompatible: input,input,1,1 1`] = `false`;

exports[`areSocketsCompatible: input,output,0,0 1`] = `true`;

exports[`areSocketsCompatible: input,output,0,1 1`] = `true`;

exports[`areSocketsCompatible: input,output,1,0 1`] = `false`;

exports[`areSocketsCompatible: input,output,1,1 1`] = `false`;

exports[`areSocketsCompatible: output,execInput,0,0 1`] = `false`;

exports[`areSocketsCompatible: output,execInput,0,1 1`] = `false`;

exports[`areSocketsCompatible: output,execInput,1,0 1`] = `false`;

exports[`areSocketsCompatible: output,execInput,1,1 1`] = `false`;

exports[`areSocketsCompatible: output,execOutput,0,0 1`] = `false`;

exports[`areSocketsCompatible: output,execOutput,0,1 1`] = `false`;

exports[`areSocketsCompatible: output,execOutput,1,0 1`] = `false`;

exports[`areSocketsCompatible: output,execOutput,1,1 1`] = `false`;

exports[`areSocketsCompatible: output,input,0,0 1`] = `true`;

exports[`areSocketsCompatible: output,input,0,1 1`] = `false`;

exports[`areSocketsCompatible: output,input,1,0 1`] = `true`;

exports[`areSocketsCompatible: output,input,1,1 1`] = `false`;

exports[`areSocketsCompatible: output,output,0,0 1`] = `false`;

exports[`areSocketsCompatible: output,output,0,1 1`] = `false`;

exports[`areSocketsCompatible: output,output,1,0 1`] = `false`;

exports[`areSocketsCompatible: output,output,1,1 1`] = `false`;

exports[`height: 0,0,0,0 1`] = `50`;

exports[`height: 0,0,0,1 1`] = `80`;

exports[`height: 0,0,0,2 1`] = `110`;

exports[`height: 0,0,1,0 1`] = `80`;

exports[`height: 0,0,1,1 1`] = `80`;

exports[`height: 0,0,1,2 1`] = `110`;

exports[`height: 0,0,2,0 1`] = `110`;

exports[`height: 0,0,2,1 1`] = `110`;

exports[`height: 0,0,2,2 1`] = `110`;

exports[`height: 0,1,0,0 1`] = `80`;

exports[`height: 0,1,0,1 1`] = `110`;

exports[`height: 0,1,0,2 1`] = `140`;

exports[`height: 0,1,1,0 1`] = `80`;

exports[`height: 0,1,1,1 1`] = `110`;

exports[`height: 0,1,1,2 1`] = `140`;

exports[`height: 0,1,2,0 1`] = `110`;

exports[`height: 0,1,2,1 1`] = `110`;

exports[`height: 0,1,2,2 1`] = `140`;

exports[`height: 0,2,0,0 1`] = `110`;

exports[`height: 0,2,0,1 1`] = `140`;

exports[`height: 0,2,0,2 1`] = `170`;

exports[`height: 0,2,1,0 1`] = `110`;

exports[`height: 0,2,1,1 1`] = `140`;

exports[`height: 0,2,1,2 1`] = `170`;

exports[`height: 0,2,2,0 1`] = `110`;

exports[`height: 0,2,2,1 1`] = `140`;

exports[`height: 0,2,2,2 1`] = `170`;

exports[`height: 1,0,0,0 1`] = `80`;

exports[`height: 1,0,0,1 1`] = `80`;

exports[`height: 1,0,0,2 1`] = `110`;

exports[`height: 1,0,1,0 1`] = `110`;

exports[`height: 1,0,1,1 1`] = `110`;

exports[`height: 1,0,1,2 1`] = `110`;

exports[`height: 1,0,2,0 1`] = `140`;

exports[`height: 1,0,2,1 1`] = `140`;

exports[`height: 1,0,2,2 1`] = `140`;

exports[`height: 1,1,0,0 1`] = `80`;

exports[`height: 1,1,0,1 1`] = `110`;

exports[`height: 1,1,0,2 1`] = `140`;

exports[`height: 1,1,1,0 1`] = `110`;

exports[`height: 1,1,1,1 1`] = `110`;

exports[`height: 1,1,1,2 1`] = `140`;

exports[`height: 1,1,2,0 1`] = `140`;

exports[`height: 1,1,2,1 1`] = `140`;

exports[`height: 1,1,2,2 1`] = `140`;

exports[`height: 1,2,0,0 1`] = `110`;

exports[`height: 1,2,0,1 1`] = `140`;

exports[`height: 1,2,0,2 1`] = `170`;

exports[`height: 1,2,1,0 1`] = `110`;

exports[`height: 1,2,1,1 1`] = `140`;

exports[`height: 1,2,1,2 1`] = `170`;

exports[`height: 1,2,2,0 1`] = `140`;

exports[`height: 1,2,2,1 1`] = `140`;

exports[`height: 1,2,2,2 1`] = `170`;

exports[`height: 2,0,0,0 1`] = `110`;

exports[`height: 2,0,0,1 1`] = `110`;

exports[`height: 2,0,0,2 1`] = `110`;

exports[`height: 2,0,1,0 1`] = `140`;

exports[`height: 2,0,1,1 1`] = `140`;

exports[`height: 2,0,1,2 1`] = `140`;

exports[`height: 2,0,2,0 1`] = `170`;

exports[`height: 2,0,2,1 1`] = `170`;

exports[`height: 2,0,2,2 1`] = `170`;

exports[`height: 2,1,0,0 1`] = `110`;

exports[`height: 2,1,0,1 1`] = `110`;

exports[`height: 2,1,0,2 1`] = `140`;

exports[`height: 2,1,1,0 1`] = `140`;

exports[`height: 2,1,1,1 1`] = `140`;

exports[`height: 2,1,1,2 1`] = `140`;

exports[`height: 2,1,2,0 1`] = `170`;

exports[`height: 2,1,2,1 1`] = `170`;

exports[`height: 2,1,2,2 1`] = `170`;

exports[`height: 2,2,0,0 1`] = `110`;

exports[`height: 2,2,0,1 1`] = `140`;

exports[`height: 2,2,0,2 1`] = `170`;

exports[`height: 2,2,1,0 1`] = `140`;

exports[`height: 2,2,1,1 1`] = `140`;

exports[`height: 2,2,1,2 1`] = `170`;

exports[`height: 2,2,2,0 1`] = `170`;

exports[`height: 2,2,2,1 1`] = `170`;

exports[`height: 2,2,2,2 1`] = `170`;

exports[`index tests: execInput,0,0,0,0 1`] = `-1`;

exports[`index tests: execInput,0,0,0,1 1`] = `-1`;

exports[`index tests: execInput,0,0,1,0 1`] = `-1`;

exports[`index tests: execInput,0,0,1,1 1`] = `-1`;

exports[`index tests: execInput,0,1,0,0 1`] = `-1`;

exports[`index tests: execInput,0,1,0,1 1`] = `-1`;

exports[`index tests: execInput,0,1,1,0 1`] = `-1`;

exports[`index tests: execInput,0,1,1,1 1`] = `-1`;

exports[`index tests: execInput,1,0,0,0 1`] = `-1`;

exports[`index tests: execInput,1,0,0,1 1`] = `-1`;

exports[`index tests: execInput,1,0,1,0 1`] = `-1`;

exports[`index tests: execInput,1,0,1,1 1`] = `-1`;

exports[`index tests: execInput,1,1,0,0 1`] = `-1`;

exports[`index tests: execInput,1,1,0,1 1`] = `-1`;

exports[`index tests: execInput,1,1,1,0 1`] = `-1`;

exports[`index tests: execInput,1,1,1,1 1`] = `-1`;

exports[`index tests: execOutput,0,0,0,0 1`] = `-1`;

exports[`index tests: execOutput,0,0,0,1 1`] = `-1`;

exports[`index tests: execOutput,0,0,1,0 1`] = `-1`;

exports[`index tests: execOutput,0,0,1,1 1`] = `-1`;

exports[`index tests: execOutput,0,1,0,0 1`] = `-1`;

exports[`index tests: execOutput,0,1,0,1 1`] = `-1`;

exports[`index tests: execOutput,0,1,1,0 1`] = `-1`;

exports[`index tests: execOutput,0,1,1,1 1`] = `-1`;

exports[`index tests: execOutput,1,0,0,0 1`] = `-1`;

exports[`index tests: execOutput,1,0,0,1 1`] = `-1`;

exports[`index tests: execOutput,1,0,1,0 1`] = `-1`;

exports[`index tests: execOutput,1,0,1,1 1`] = `-1`;

exports[`index tests: execOutput,1,1,0,0 1`] = `-1`;

exports[`index tests: execOutput,1,1,0,1 1`] = `-1`;

exports[`index tests: execOutput,1,1,1,0 1`] = `-1`;

exports[`index tests: execOutput,1,1,1,1 1`] = `-1`;

exports[`index tests: input,0,0,0,0 1`] = `-1`;

exports[`index tests: input,0,0,0,1 1`] = `-1`;

exports[`index tests: input,0,0,1,0 1`] = `0`;

exports[`index tests: input,0,0,1,1 1`] = `0`;

exports[`index tests: input,0,1,0,0 1`] = `-1`;

exports[`index tests: input,0,1,0,1 1`] = `-1`;

exports[`index tests: input,0,1,1,0 1`] = `0`;

exports[`index tests: input,0,1,1,1 1`] = `0`;

exports[`index tests: input,1,0,0,0 1`] = `-1`;

exports[`index tests: input,1,0,0,1 1`] = `-1`;

exports[`index tests: input,1,0,1,0 1`] = `0`;

exports[`index tests: input,1,0,1,1 1`] = `0`;

exports[`index tests: input,1,1,0,0 1`] = `-1`;

exports[`index tests: input,1,1,0,1 1`] = `-1`;

exports[`index tests: input,1,1,1,0 1`] = `0`;

exports[`index tests: input,1,1,1,1 1`] = `0`;

exports[`index tests: output,0,0,0,0 1`] = `-1`;

exports[`index tests: output,0,0,0,1 1`] = `0`;

exports[`index tests: output,0,0,1,0 1`] = `-1`;

exports[`index tests: output,0,0,1,1 1`] = `0`;

exports[`index tests: output,0,1,0,0 1`] = `-1`;

exports[`index tests: output,0,1,0,1 1`] = `0`;

exports[`index tests: output,0,1,1,0 1`] = `-1`;

exports[`index tests: output,0,1,1,1 1`] = `0`;

exports[`index tests: output,1,0,0,0 1`] = `-1`;

exports[`index tests: output,1,0,0,1 1`] = `0`;

exports[`index tests: output,1,0,1,0 1`] = `-1`;

exports[`index tests: output,1,0,1,1 1`] = `0`;

exports[`index tests: output,1,1,0,0 1`] = `-1`;

exports[`index tests: output,1,1,0,1 1`] = `0`;

exports[`index tests: output,1,1,1,0 1`] = `-1`;

exports[`index tests: output,1,1,1,1 1`] = `0`;

exports[`move 1`] = `
Object {
  "_id": "test",
  "code": "cbId",
  "mstPouchType": "Box",
  "name": "hal",
  "sockets": Array [],
  "values": Array [],
  "x": 1,
  "y": 2,
}
`;

exports[`move 2`] = `
Object {
  "_id": "test",
  "code": "cbId",
  "mstPouchType": "Box",
  "name": "hal",
  "sockets": Array [],
  "values": Array [],
  "x": 2,
  "y": 4,
}
`;

exports[`x/y execInput 1`] = `
Object {
  "x": 18,
  "y": 50,
}
`;

exports[`x/y execInput 2`] = `
Object {
  "x": -82,
  "y": 150,
}
`;

exports[`x/y execInput 3`] = `
Object {
  "x": -82,
  "y": 210,
}
`;

exports[`x/y execInput 4`] = `
Object {
  "x": 18,
  "y": 240,
}
`;

exports[`x/y execInput 5`] = `
Object {
  "x": -482,
  "y": 200,
}
`;

exports[`x/y execOutput 1`] = `
Object {
  "x": -18,
  "y": 50,
}
`;

exports[`x/y execOutput 2`] = `
Object {
  "x": -68,
  "y": 150,
}
`;

exports[`x/y execOutput 3`] = `
Object {
  "x": -68,
  "y": 210,
}
`;

exports[`x/y execOutput 4`] = `
Object {
  "x": 2,
  "y": 240,
}
`;

exports[`x/y execOutput 5`] = `
Object {
  "x": -418,
  "y": 200,
}
`;

exports[`x/y input 1`] = `
Object {
  "x": 18,
  "y": 50,
}
`;

exports[`x/y input 2`] = `
Object {
  "x": -82,
  "y": 150,
}
`;

exports[`x/y input 3`] = `
Object {
  "x": -82,
  "y": 210,
}
`;

exports[`x/y input 4`] = `
Object {
  "x": 18,
  "y": 240,
}
`;

exports[`x/y input 5`] = `
Object {
  "x": -482,
  "y": 200,
}
`;

exports[`x/y output 1`] = `
Object {
  "x": -18,
  "y": 50,
}
`;

exports[`x/y output 2`] = `
Object {
  "x": -68,
  "y": 150,
}
`;

exports[`x/y output 3`] = `
Object {
  "x": -68,
  "y": 210,
}
`;

exports[`x/y output 4`] = `
Object {
  "x": 2,
  "y": 240,
}
`;

exports[`x/y output 5`] = `
Object {
  "x": -418,
  "y": 200,
}
`;
